### ISSUE TITLE
mysql grammar is three times slower with grouped parsing than individual parsing!

### ISSUE THREAD

**Issue by kaby76** [antlr/antlr4#4324]
This is a problem that seems to occur across targets for the [MySQL grammar](https://github.com/antlr/grammars-v4/tree/master/sql/mysql/Positive-Technologies), but it's problematic for the Go target. The Go target is much slower than CSharp or Java. The testing fails for Go because it hits the 300s kill watchdog on the Mac. (The watchdog prevents tests hanging and requiring a manual kill of the Github Action.)

"Grouped parsing" is a parse of multiple tests using one program, where each parse creates a lexer/parser object in one process space. Antlr is designed to share cached information across parses even though the lexer/parser objects are not shared across parses. "Warm up", a term used by some, it an informal and undefined term for "grouped parsing". "Individual parsing" is a parse of one test per program. In this scenario, there is no cached information to improve parsing. (We should be using formally defined terms.)

For the MySQL grammar, grouped parsing does exactly the opposite of what we would expect: it slows down the parse. I suspect that it is caused because the MySQL grammar is ambiguous. https://github.com/antlr/grammars-v4/pull/3545#issuecomment-1600544235

**Comment by jimidle** [comment]
Yes - I decided not to put very much more - maybe no - effort in to grammars that are this bad. It is faster than it was, but I am not sure my effort is best spent on this. I have some potential improvements that will probably help this a bit, but they need serious time to check out. 

The real issue is that that parse is not usable. I started to rework it, but there is already a much better mysql parser out there I am told. I think that this parser is actually a target for removal unless the original author(s) wish to fix it.

I can look at this again  though - do you have the Go code that shows what you mean by grouped parsing though? My tests either use a new lexer/parser for each parse, or reuse the lexer/parser for each example that comes with that parser. When I reuse the parser, I do (as of this release anyway ;), get a big improvement locally. Your comments seem to talking about something else? Or maybe I just misread them?

The parser basically causes GB of allocations - to the point where I wondered about it being faster not to use the cache for this grammar. But then I am putting work into helping bad grammars and that seems counterproductive when reasonable-to-good grammars are now very fast (and will get faster).
